query_execution.go

```diff
@@ -0,0 +1,889 @@
+	InsertionPoint []string
+	Data           interface{}
+	Errors         gqlerror.List
+}
+
+type queryExecution struct {
+	ctx            context.Context
+	schema         *ast.Schema
+	requestCount   int32
+	maxRequest     int32
+	graphqlClient  *GraphQLClient
+	boundaryFields BoundaryFieldsMap
+
+	group   *errgroup.Group
+	results chan executionResult
+}
+
+func newQueryExecution(ctx context.Context, client *GraphQLClient, schema *ast.Schema, boundaryFields BoundaryFieldsMap, maxRequest int32) *queryExecution {
+	group, ctx := errgroup.WithContext(ctx)
+	return &queryExecution{
+		ctx:            ctx,
+		schema:         schema,
+		graphqlClient:  client,
+		boundaryFields: boundaryFields,
+		maxRequest:     maxRequest,
+		group:          group,
+		results:        make(chan executionResult),
+	}
+}
+
+func (q *queryExecution) Execute(queryPlan *QueryPlan) ([]executionResult, gqlerror.List) {
+	readWg := &sync.WaitGroup{}
+	results := []executionResult{}
+
+	if len(queryPlan.RootSteps) > int(q.maxRequest) {
+		return nil, gqlerror.List{
+			&gqlerror.Error{
+				Message: fmt.Sprintf("exceeded max requests of %v", q.maxRequest),
+			},
+		}
+	}
+
+	for _, step := range queryPlan.RootSteps {
+		if step.ServiceURL == internalServiceName {
+			r, err := executeBrambleStep(step)
+			if err != nil {
+				return nil, q.createGQLErrors(step, err)
+			}
+			results = append(results, *r)
+			continue
+		}
+
+		step := step
+		q.group.Go(func() error {
+			return q.executeRootStep(step)
+		})
+	}
+
+	readWg.Add(1)
+	go func() {
+		for result := range q.results {
+			results = append(results, result)
+		}
+		readWg.Done()
+	}()
+
+	if err := q.group.Wait(); err != nil {
+		return nil, gqlerror.List{
+			&gqlerror.Error{
+				Message: err.Error(),
+			},
+		}
+	}
+	close(q.results)
+	readWg.Wait()
+	return results, nil
+}
+
+func (q *queryExecution) executeRootStep(step *QueryPlanStep) error {
+	var document string
+
+	switch operationType := step.ParentType; operationType {
+	case queryObjectName, mutationObjectName:
+		document = strings.ToLower(operationType) + formatSelectionSet(q.ctx, q.schema, step.SelectionSet)
+	default:
+		return errors.New("expected mutation or query root step")
+	}
+
+	var data map[string]interface{}
+
+	err := q.executeDocument(document, step.ServiceURL, &data)
+	if err != nil {
+		q.writeExecutionResult(step, data, err)
+		return nil
+	}
+
+	q.writeExecutionResult(step, data, nil)
+
+	for _, childStep := range step.Then {
+		boundaryIDs, err := extractAndDedupeBoundaryIDs(data, childStep.InsertionPoint)
+		if err != nil {
+			return err
+		}
+		if len(boundaryIDs) == 0 {
+			continue
+		}
+
+		childStep := childStep
+		q.group.Go(func() error {
+			return q.executeChildStep(childStep, boundaryIDs)
+		})
+	}
+	return nil
+}
+
+func (q *queryExecution) executeDocument(document string, serviceURL string, response interface{}) error {
+	req := NewRequest(document).
+		WithHeaders(GetOutgoingRequestHeadersFromContext(q.ctx))
```

Comment:
Is this helper function worth it just to inline one assignment? It hides the modification side effect.

Reply:
Not a fan of mutating the request after initialisation regardless but I agree this doesn't really help in that area. However setting public fields directly on the struct after initialisation felt more wrong to me than using a helper function, ideally the `Headers` field wouldn't be public at all but it's a little hard to change now. I don't feel too strongly either way though, my preference would to completely redo the initialisation at some point.